execution/executor/src/chunk_executor.rs

```diff
@@ -264,189 +254,71 @@ impl<V: VMExecutor> ChunkExecutorInner<V> {
-                .map(|t| t.into())
-                .collect::<Vec<_>>()
-        });
-
-        // Execute transactions.
-        let state_view = self.latest_state_view(&parent_state)?;
-        let chunk_output = {
-            let _timer = VM_EXECUTE_CHUNK.start_timer();
-            // State sync executor shouldn't have block gas limit.
-            ChunkOutput::by_transaction_execution::<V>(
-                sig_verified_txns.into(),
-                state_view,
-                BlockExecutorConfigFromOnchain::new_no_block_limit(),
-            )?
-        };
-
-        // Calcualte state snapshot
-        let (result_state, next_epoch_state, state_checkpoint_output) =
-            ApplyChunkOutput::calculate_state_checkpoint(
-                chunk_output,
-                &self.commit_queue.lock().latest_state(),
-                None, // append_state_checkpoint_to_block
-                Some(known_state_checkpoints),
-                false, // is_block
-            )?;
-
-        // Enqueue for next stage.
-        self.commit_queue
-            .lock()
-            .enqueue_for_ledger_update(ChunkToUpdateLedger {
-                result_state,
-                state_checkpoint_output,
-                next_epoch_state,
-                verified_target_li,
-                epoch_change_li,
-                txn_infos_with_proof,
-            })?;
-
-        info!(
-            LogSchema::new(LogEntry::ChunkExecutor)
-                .first_version_in_request(Some(first_version_in_request))
-                .num_txns_in_request(num_txns),
-            "Executed transaction chunk!",
-        );
-
-        Ok(())
+        THREAD_MANAGER
+            .get_exe_cpu_pool()
+            .install(|| chunk.verify_chunk(ledger_info, first_version))
     }
 
-    fn enqueue_chunk_by_transaction_outputs(
+    // ************************* Chunk Executor Implementation *************************
+    fn enqueue_chunk<Chunk: TransactionChunkWithProof + Sync>(
         &self,
-        txn_output_list_with_proof: TransactionOutputListWithProof,
+        chunk: Chunk,
         verified_target_li: &LedgerInfoWithSignatures,
         epoch_change_li: Option<&LedgerInfoWithSignatures>,
+        mode_for_log: &'static str,
     ) -> Result<()> {
-        let _timer = APPLY_CHUNK.start_timer();
-
-        let num_txns = txn_output_list_with_proof.transactions_and_outputs.len();
-        ensure!(num_txns != 0, "Empty transaction list!");
-        let first_version_in_request = txn_output_list_with_proof
-            .first_transaction_output_version
-            .ok_or_else(|| anyhow!("Non-empty chunk with first_version == None."))?;
         let parent_state = self.commit_queue.lock().latest_state();
-        ensure!(
-            first_version_in_request == parent_state.next_version(),
-            "Unexpected chunk. version in request: {}, current_version: {:?}",
-            first_version_in_request,
-            parent_state.current_version,
-        );
```

Comment:
These are guaranteed if list.verify() passes with Some(expected_version) passed in, specifically, it's guranteed that 1. the list is not empty and 2. the list starts from expected_version